New anti-pattern (#12997)
elixir-lang/elixir: lib/elixir/pages/anti-patterns/code-anti-patterns.md

  * If you are the maintainer for both `plug` and `plug_auth`, then you may allow `plug_auth` to define modules with the `Plug` namespace, such as `Plug.Auth`. However, you are responsible for avoiding or managing any conflicts that may arise in the future

## Non-assertive patterns

#### Problem

> Overall, Elixir systems are composed of many supervised processes, so the effects of an error are localized to a single process, not propagating to the entire application. A supervisor will detect the failing process, report it, and possibly restart it. This anti-pattern arises when developers write defensive or imprecise code, capable of returning incorrect values which were not planned for, instead of programming in an assertive style through pattern matching and guards.

Reviewer: I don't think the future is needed here?

```suggestion
Overall, Elixir systems are composed of many supervised processes, so the effects of an error are localized to a single process, and don't propagate to the entire application. A supervisor detects the failing process, reports it, and possibly restarts it. This anti-pattern arises when developers write defensive or imprecise code, capable of returning incorrect values which were not planned for, instead of programming in an assertive style through pattern matching and guards.
```

The function `get_value/2` tries to extract a value from a specific key of a URL query string. As it is not implemented using pattern matching, `get_value/2` always returns a value, regardless of the format of the URL query string passed as a parameter in the call. Sometimes the returned value will be valid. However, if a URL query string with an unexpected format is used in the call, `get_value/2` will extract incorrect values from it:

```elixir
defmodule Extract do
  def get_value(string, desired_key) do
    parts = String.split(string, "&")

    Enum.find_value(parts, fn pair ->
      key_value = String.split(pair, "=")
      Enum.at(key_value, 0) == desired_key && Enum.at(key_value, 1)
    end)
  end
end
```

```elixir
# URL query string with the planned format - OK!
iex> Extract.get_value("name=Lucas&university=UFMG&lab=ASERG", "lab")
"ASERG"
iex> Extract.get_value("name=Lucas&university=UFMG&lab=ASERG", "university")
"UFMG"
# Unplanned URL query string format - Unplanned value extraction!
iex> Extract.get_value("name=Lucas&university=institution=UFMG&lab=ASERG", "university")
"institution"   # <= why not "institution=UFMG"? or only "UFMG"?
```

#### Refactoring

> To remove this anti-pattern, `get_value/2` can be refactored through the use of pattern matching. So, if an unexpected URL query string format is used, the function will crash instead of returning an invalid value. This behaviour, shown below, will allow clients to decide how to handle these errors and will not give a false impression that the code is working correctly when unexpected values are extracted:

Reviewer: I don't think the future is needed, as above:

```suggestion
To remove this anti-pattern, `get_value/2` can be refactored through the use of pattern matching. So, if an unexpected URL query string format is used, the function will crash instead of returning an invalid value. This behaviour, shown below, allows clients to decide how to handle these errors and doesn't give a false impression that the code is working correctly when unexpected values are extracted:
```

    Enum.find_value(parts, fn pair ->
      [key, value] = String.split(pair, "=") # <= pattern matching
      key == desired_key && value
    end)
  end
end
```

```elixir
# URL query string with the planned format - OK!
iex> Extract.get_value("name=Lucas&university=UFMG&lab=ASERG", "name")
"Lucas"
# Unplanned URL query string format - Crash explaining the problem to the client!
iex> Extract.get_value("name=Lucas&university=institution=UFMG&lab=ASERG", "university")
** (MatchError) no match of right hand side value: ["university", "institution", "UFMG"]
  extract.ex:7: anonymous fn/2 in Extract.get_value/2 # <= left hand: [key, value] pair
iex> Extract.get_value("name=Lucas&university&lab=ASERG", "university")
** (MatchError) no match of right hand side value: ["university"]
  extract.ex:7: anonymous fn/2 in Extract.get_value/2 # <= left hand: [key, value] pair
```

Elixir and pattern matching promote an assertive style of programming where you handle the known cases. Once an unexpected scenario arises, you can decide to address it accordingly based on practical examples, or conclude the scenario is indeed invalid and the exception is the desired choice.

> `case/2` is another important construct in Elixir that help us write assertive code, by matching on specific patterns. For exmaple, if a function returns `{:ok, ...}` or `{:error, ...}`, prefer to explicitly match on both patterns:

Reviewer: Typo:

```suggestion
`case/2` is another important construct in Elixir that help us write assertive code, by matching on specific patterns. For example, if a function returns `{:ok, ...}` or `{:error, ...}`, prefer to explicitly match on both patterns:
```

```elixir
case some_function(arg) do
  {:ok, value} -> # ...
  {:error, _} -> # ...
end
```

In particular, avoid matching solely on `_`, as shown below, as it is less clear in intent and it may hide bugs if `some_function/1` adds new return values in the future:

```elixir
case some_function(arg) do
  {:ok, value} -> # ...
  _ -> # ...
end
```

> ## Non-assertive truthiness

Reviewer 2: ```suggestion
#### Additional remarks

This anti-pattern was formerly known as [Speculative assumptions](https://github.com/lucasvegi/Elixir-Code-Smells#speculative-assumptions).

## Non-assertive truthiness
```

#### Problem

> Elixir provides the concept of truthiness: where `nil` and `false` are considered "falsy" and all other values are "truthy". Many constructs in the language, such as `&&/2`, `||/2`, and `!/1` handle truthy and falsy values. Using those operators is not an anti-pattern. However, using those operators when all operands are expected to be booleans, may be an anti-pattern.

Reviewer: Maybe quotes make this nicer:

```suggestion
Elixir provides the concept of "truthiness": `nil` and `false` are considered "falsy" and all other values are "truthy". Many constructs in the language, such as `&&/2`, `||/2`, and `!/1` handle truthy and falsy values. Using those operators is not an anti-pattern. However, using those operators when all operands are expected to be booleans, may be an anti-pattern.
```

#### Example

The simplest scenario where this anti-pattern manifests is in conditionals, such as:

```elixir
if is_binary(name) && is_integer(age) do
  # ...
else
  # ...
end
```

Given both operands of `&&/2` are booleans, the code is more generic than necessary, and potentially unclear.

#### Refactoring

> To remove this anti-pattern, we can replace `&&/2`, `||/2`, and `!/1` by `and/2`, `or/2`, and `not/1` respectively. The new operators assert at least their first argument is a boolean:

Reviewer: ```suggestion
To remove this anti-pattern, we can replace `&&/2`, `||/2`, and `!/1` by `and/2`, `or/2`, and `not/1` respectively. These operators assert at least their first argument is a boolean:
```

```elixir
if is_binary(name) or is_integer(age) do
  # ...
else
  # ...
end
```

This technique may be particularly important when working with Erlang code. Erlang does not have the concept of truthiness. It never returns `nil`, instead its functions may return `:error` or `:undefined` in places an Elixir developer would return `nil`. Therefore, to avoid accidentally interpreting `:undefined` or `:error` as a truthy value, you may prefer to use `and/2`, `or/2`, and `not/1` exclusively when interfacing with Erlang APIs.

> ## Non-existent map keys

Author: ```suggestion
## Non-assertive map access
```

`case/2` is another important construct in Elixir that help us write assertive code, by matching on specific patterns. For exmaple, if a function returns `{:ok, ...}` or `{:error, ...}`, prefer to explicitly match on both patterns:

```elixir
case some_function(arg) do
  {:ok, value} -> # ...
  {:error, _} -> # ...
end
```

In particular, avoid matching solely on `_`, as shown below, as it is less clear in intent and it may hide bugs if `some_function/1` adds new return values in the future:

```elixir
case some_function(arg) do
  {:ok, value} -> # ...
  _ -> # ...
end
```

## Non-assertive truthiness

#### Problem

Elixir provides the concept of truthiness: where `nil` and `false` are considered "falsy" and all other values are "truthy". Many constructs in the language, such as `&&/2`, `||/2`, and `!/1` handle truthy and falsy values. Using those operators is not an anti-pattern. However, using those operators when all operands are expected to be booleans, may be an anti-pattern.

#### Example

> The simplest scenario this anti-pattern manifest is conditions, such as:

Author: ```suggestion
The simplest scenario where this anti-pattern manifests is in conditionals, such as:
```

  * If you are the maintainer for both `plug` and `plug_auth`, then you may allow `plug_auth` to define modules with the `Plug` namespace, such as `Plug.Auth`. However, you are responsible for avoiding or managing any conflicts that may arise in the future

> ## Non-assertive patterns

Reviewer 3: The word "pattern" might be slightly confusing due to "pattern-matching" (which in this case it's precisely not).
How about `Non-assertive code` or `Non-assertive style` (both inspired from this [article](https://dashbit.co/blog/writing-assertive-code-with-elixir))?

Author: I was playing intentionally on the ambiguity. Although I would say, even in the poor case, it is pattern matching, because `foo = ...` is a pattern matching (but one with no assertion).

Author: I would be happy to rename this to `Non-assertive pattern matching` though.

Reviewer 3: Makes sense, I agree that `Non-assertive pattern matching` would clear up that ambiguity if this is the intention.

Reviewer: Agreed, *Non-assertive pattern matching* sounds great.

Author: ```suggestion
## Non-assertive pattern matching
```

#### Problem

> Overall, Elixir systems are composed of many supervised processes, so the effects of an error are localized to a single process, not propagating to the entire application. A supervisor will detect the failing process, report it, and possibly restart it. This means Elixir developers can write code in an assertive style, instead of programming defensively and returning incorrect values which were not planned for.

Author: ```suggestion
Overall, Elixir systems are composed of many supervised processes, so the effects of an error are localized to a single process, not propagating to the entire application. A supervisor will detect the failing process, report it, and possibly restart it. This anti-pattern arises when developers write defensive or imprecise code, capable of returning incorrect values which were not planned for, instead of programming in an assertive style through pattern matching and guards.
```

```elixir
defmodule Extract do
  def get_value(string, desired_key) do
    parts = String.split(string, "&")

    Enum.find_value(parts, fn pair ->
      [key, value] = String.split(pair, "=") # <= pattern matching
      key == desired_key && value
    end)
  end
end
```

```elixir
# URL query string with the planned format - OK!
iex> Extract.get_value("name=Lucas&university=UFMG&lab=ASERG", "name")
"Lucas"
# Unplanned URL query string format - Crash explaining the problem to the client!
iex> Extract.get_value("name=Lucas&university=institution=UFMG&lab=ASERG", "university")
** (MatchError) no match of right hand side value: ["university", "institution", "UFMG"]
  extract.ex:7: anonymous fn/2 in Extract.get_value/2 # <= left hand: [key, value] pair
iex> Extract.get_value("name=Lucas&university&lab=ASERG", "university")
** (MatchError) no match of right hand side value: ["university"]
  extract.ex:7: anonymous fn/2 in Extract.get_value/2 # <= left hand: [key, value] pair
```

> Elixir and pattern matching promote an assertive style of programming where you handle the known cases. Once an unexpected scenario arises, you can decide to address it accordingly based on practical examples, or conclude the scenario is indeed invalid and the exception is the desired choice.

Author: This paragraph until the end of the diff is new.

`case/2` is another important construct in Elixir that help us write assertive code, by matching on specific patterns. For exmaple, if a function returns `{:ok, ...}` or `{:error, ...}`, prefer to explicitly match on both patterns:

> 

Author: We already have a separate code smell for that (non-existing map keys). Which perhaps could be retitled to "non-assertive map access".

> 

Reviewer 3: I like it!

> 

Reviewer 2: @josevalim, I agree with your proposed changes to these names. However, I suggest that whenever name changes are made, we include a note at the end of the documentation for the renamed smell, stating that the name has changed compared to the research, and provide a link to the original repository. We have already done this with "Scattered process interfaces" (originally "Agent obsession").

My suggestion is to preserve the research history and enable readers of the Elixir official documentation to correlate our prevalence and relevance data for smells, as outlined in our EMSE paper, with the new names in the documentation.

#### Example

The simplest scenario this anti-pattern manifest is conditions, such as:

```elixir
if is_binary(name) && is_integer(age) do
  # ...
else
  # ...
end
```

Given both operands of `&&/2` are booleans, the code is more generic than necessary, and potentially unclear.

#### Refactoring

To remove this anti-pattern, we can replace `&&/2`, `||/2`, and `!/1` by `and/2`, `or/2`, and `not/1` respectively. The new operators assert at least their first argument is a boolean:

```elixir
if is_binary(name) or is_integer(age) do
  # ...
else
  # ...
end
```

A more assertive version of the code may also replace `if/2` by `case/2`, applying similar concepts to the ones presented in [Non-assertive patterns](#non-assertive-patterns):

```elixir
case is_binary(name) or is_integer(age) do
  true -> # ...
  false -> # ...
end

> ```

Reviewer 3: Is it really something we should be promoting?

I don't really have an opinion, but I thought this might be controversial and I'm not even sure this style is used in the elixir codebase.
(also interestingly elixir-styler took the opposite choice of [rewriting these boolean `case`s as `if`s](https://github.com/adobe/elixir-styler/blob/main/test/style/single_node_test.exs#L295-L310))

> A more assertive version of the code may also replace `if/2` by `case/2`, applying similar concepts to the ones presented in [Non-assertive patterns](#non-assertive-patterns):
> 
> ```elixir
> case is_binary(name) or is_integer(age) do
>   true -> # ...
>   false -> # ...
> end
> ```
> 
> These techniques may be particularly important when working with Erlang code. Erlang does not have the concept of truthiness. It never returns `nil`, instead its functions may return `:error` or `:undefined` in places an Elixir developer would return `nil`.

Author: ````suggestion
This technique may be particularly important when working with Erlang code. Erlang does not have the concept of truthiness. It never returns `nil`, instead its functions may return `:error` or `:undefined` in places an Elixir developer would return `nil`. Therefore, to avoid accidentally interpreting `:undefined` or `:error` as a truthy value, it may be preferable to use `and/2`, `or/2`, and `not/1` exclusively when interfacing with Erlang APIs.
````

> 
> Therefore, to avoid accidentally interpreting `:undefined` or `:error` as a truthy value, it may be preferrable to use `and/2`, `or/2`, `not/1`, and `case/2` exclusively when interfacing with Erlang APIs.

Author: ```suggestion
```

iex> point[:x] # <= by default, struct does not support dynamic access
** (UndefinedFunctionError) ... (Point does not implement the Access behaviour)

> ```

Reviewer 2: Another suggestion, just to maintain the research history in cases of renaming:

```suggestion
```

#### Additional remarks
This anti-pattern was formerly known as [Accessing non-existent map/struct fields](https://github.com/lucasvegi/Elixir-Code-Smells#accessing-non-existent-mapstruct-fields).
```

> ```

Reviewer 2: Another suggestion, just to maintain the research history in cases of renaming:

```suggestion
```

#### Additional remarks
This anti-pattern was formerly known as [Accessing non-existent map/struct fields](https://github.com/lucasvegi/Elixir-Code-Smells#accessing-non-existent-mapstruct-fields).
```
